### PR TITLE
Initial CRUD/API functions

### PR DESCRIPTION
**Pull request by rfmeier**
## Description

This pull request adds the initial CRUD/API functions. These functions can assist with external plugins working with Content Model entries.

### insert_entry_model()
```
insert_model_entry( string $model_slug, array $field_data, array $post_data = [] )
```

### update_model_entry()
```
update_model_entry( int $post_id, array $field_data, array $post_data = [] )
```

### add_relationship()
```
add_relationship( int $post_id, string $relationship_field_slug, int $relationship_id )
```

### replace_relationship()
```
replace_relationship( int $post_id, string $relationship_field_slug, array $relationship_ids )
```

### get_relationship()
```
get_relationship( int $post_id, string $relationship_field_slug )
```

### fetch_model_field()
```
fetch_model_field( string $model, string $field_slug ): ?array 
```

### fetch_model()
```
fetch_model( string $model ): ?array
```

## Jira
https://wpengine.atlassian.net/browse/MTKA-1443
https://wpengine.atlassian.net/browse/MTKA-1445

Documentation to follow.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,15 @@
 - Focus now moves to the new input field when adding a new repeating number, date or single line text field row.
 - Pressing enter in a repeating number, date or single line text field now moves focus to the next field, or adds a new field if focus is in the last field.
 
+### Added
+- Function insert_entry_model() for inserting an model entry and fields.
+- Function update_model_entry() for updating and existing entry model and fields.
```
Comment (nickcernis) [inline CHANGELOG.md:10]: nit: typos

```suggestion
- Function insert_entry_model() for inserting a model entry and fields.
- Function update_model_entry() for updating an existing entry model and fields.
```

File: includes/api/crud-functions.php [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace WPE\AtlasContentModeler\API;
+
+use function WPE\AtlasContentModeler\get_field_from_slug;
+use function WPE\AtlasContentModeler\sanitize_fields;
+use function WPE\AtlasContentModeler\ContentRegistration\get_registered_content_types;
+use function WPE\AtlasContentModeler\API\Utility\get_data_for_fields;
```
Comment (nickcernis) [inline includes/api/crud-functions.php:15]: Looks like `get_data_for_fields` is unused in this file.

File: includes/api/crud-functions.php [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+ *
+ * @param int          $post_id         The id of the post being updated.
+ * @param array        $field_data      The field data being saved.
+ * @param array        $post_data       The post data object.
+ * @param bool|boolean $skip_validation Option to skip validation.
+ *
+ * @return int|WP_Error The updated content model entry id or WP_Error.
+ */
+function update_model_entry( int $post_id, array $field_data, array $post_data = [], bool $skip_validation = false ) {
+	$post_data = get_post( $post_id );
```
Comment (nickcernis) [inline includes/api/crud-functions.php:79]: Should this be named something else, such as `$original_post_data` (with some extra logic to merge it with the passed in `$post_data`)? 

As spotted by @mindctrl, the function accepts a `$post_data` parameter, presumably to provide optional post data to update separately from the field data, but we're immediately overriding it here with the original data from `get_post`, so post data from the user gets ignored.

The tests in `test-crud-functions.php` always pass an empty array for as the third argument, so we don't validate that post data can be updated separately to field data.

File: tests/integration/api/test-crud-functions.php [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+		update_registered_content_types( $this->content_models );
+
+		$model_id = insert_model_entry( 'validation', $data );
+
+		$wp_post = get_post( $model_id );
+		$this->assertEquals( '', $wp_post->post_title );
+	}
+
+	public function test_update_model_entry_will_update_post_meta_and_return_post_id_on_success() {
+		$model_schema = $this->content_models['validation'];
```
Comment (nickcernis) [inline tests/integration/api/test-crud-functions.php:107]: `$model_schema` appears to be unused in this test.

File: includes/api/crud-functions.php (hunk 2 above)
Comment (rfmeier) [reply includes/api/crud-functions.php:15]: Ah yes. Good catch. Thank you.

Comment (rfmeier) [reply includes/api/crud-functions.php:15]: The function `array_get_key_value()` also existed, but wasn't being used. Removed this.

File: includes/api/crud-functions.php (hunk 3 above)
Comment (rfmeier) [reply includes/api/crud-functions.php:79]: Now I see it `$post_data = get_post( $post_id );`. Yeah-- This is definitely overriding. I don't think it was intentional.

File: CHANGELOG.md (hunk 1 above)
Comment (rfmeier) [reply CHANGELOG.md]: Thanks.

**Review by nickcernis — COMMENTED** [review]
Great work, @anthonyburchell and @rfmeier!

Dropped a couple of notes above, but this is looking really good. I made a quick plugin in `wp-content/plugins/test.php` to try it out as WP plugin devs might.

```php
<?php
/* Plugin Name: Test */

namespace WPE\Test;

use function WPE\AtlasContentModeler\API\insert_model_entry;

add_action( 'wp_loaded', __NAMESPACE__ . '\test' );

function test() {
	// error_log( 'inserting bunny' );
	$field_data = [
		'name' => 'Bunny from the API',
	];

	// $field_data = [];
	
	$post_data = [
		'post_status' => 'publish',
	];	
	
	$skip_validation = false;

	$new_entry = insert_model_entry(
		'rabbit',
		$field_data,
		$post_data,
		$skip_validation		
	);
		
	// error_log( print_r( $new_entry, true ) );	
}
```

The rabbit factory works smoothly on page refresh. 🐰  🐰  🐰 

Screenshot 2022-04-20 at 16 12 47

And returns errors if I pass `[]` as `$field_data`:

```
[20-Apr-2022 14:25:36 UTC] WP_Error Object
(
    [errors] => Array
        (
            [invalid_model_field] => Array
                (
                    [0] => Name field is required
                )

        )

    [error_data] => Array
        (
        )

    [additional_data:protected] => Array
        (
        )
)
```

File: includes/api/crud-functions.php (hunk 3 above)
Comment (nickcernis) [reply includes/api/crud-functions.php]: Nice, looks fixed as of [3a1d6d1](https://github.com/wpengine/atlas-content-modeler/pull/503/commits/3a1d6d12516c3a50abe8bd59af3d2a4f06d084ef).

**Review by nickcernis — DISMISSED** [review]

File: includes/api/validation-functions.php [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+ *
+ * @return void
+ */
+function validate_number_field( $value, array $field ): void {
+	if ( is_field_required( $field ) ) {
+		validate_not_empty( $value, "{$field['name']} cannot be empty" );
+	}
+
+	if ( is_field_repeatable( $field ) ) {
+		validate_array( $value, "{$field['name']} must be an array of {$field['type']}" );
```
Comment (mindctrl) [inline includes/api/validation-functions.php:128]: This verifies that the value is an array, but doesn't verify that the array values are numbers. You can pass an array of strings and it validates fine.

Comment (rfmeier) [reply includes/api/validation-functions.php:128]: That is correct. During discussion it was agreed that we wouldn't validate the contents in the array of a repeatable field value... yet. There is currently a ticket to validate the contents of a repeatable field.

**Review by mindctrl — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 6]
```diff
@@ -5,6 +5,15 @@
...
 - Pressing enter in a repeating number, date or single line text field now moves focus to the next field, or adds a new field if focus is in the last field.
 
+### Added
+- Function insert_entry_model() for inserting a model entry and fields.
+- Function update_model_entry() for updating an existing entry model and fields.
+- Function replace_relationship() to replace/add model entry relationships.
+- Function add_relationship() to append model entry relationships.
+- Function get_relationship() to retrieve the relationship object.
+- Function fetch_model() to retrieve the model schema.
+- Function fetch_model_field() to retrieve the model field schema.
```
[the remark this answers is not in the document]
Comment (rfmeier) [reply CHANGELOG.md:15]: Consistency is king. @anthonyburchell any thoughts about replacing `fetch_` with `get_` ?

File: includes/publisher/lib/field-functions.php [hunk 7]
```diff
@@ -91,6 +91,53 @@ function is_field_featured_image( string $slug, array $fields ): bool {
...
+		case 'date':
+			$key = 'isRepeatableDate';
+			break;
+		case 'media':
+			$key = 'isRepeatableMedia';
+			break;
+	}
+
+	if ( $key && \array_key_exists( $key, $field ) ) {
+		$repeatable = ( true === $field[ $key ] || 'true' === $field[ $key ] );
```
[the remark this answers is not in the document]
Comment (rfmeier) [reply includes/publisher/lib/field-functions.php:135]: >Have we mixed those in our schema data?

Yes. It has been inconsistent.

[the remark this answers is not in the document]
Comment (rfmeier) [reply includes/publisher/lib/field-functions.php:135]: I was seeing values within field data arrays where there is a mix of `'true'` and `true`. This was to make up for that inconsistency.

File: CHANGELOG.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (anthonyburchell) [reply CHANGELOG.md]: agreed. 👍

**Review by mindctrl — APPROVED** [review]

**Pull request merged**